Links within links: Take 2

SiAdcock opened the pull request:
Attempt number 2 at removing the overlays (attempt 1: #18470) 

## Changes

1. [`6668880`](https://github.com/guardian/frontend/pull/18509/commits/6668880b56e96fc53e9ab5ff9f4e784fc8e5639c) Prevents layout thrashing by setting default height and widths for the `<object>` element
2. [`5228e5e`](https://github.com/guardian/frontend/pull/18509/commits/5228e5e027c37bc612bf68a8de0914f9c56c23da) Sets a default colour for `fc-item` elements, which will be inherited by the headline for most tones
3. [`d76c306`](https://github.com/guardian/frontend/pull/18509/commits/d76c3060c8519ea4e0924f07c343a372df4a3c6a) Fixes rich link styling in a similar manner to the above

- [x] Test in CODE

PRBuilds commented:
PRbuilds results:


**Screenshots**
[desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png)



&#x1F49A; **Exceptions**
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/thrown-exceptions.js)





&#x1F49A; **A11y validation**
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/a11y-report.txt)



**Apache Benchmark Load Testing**
[loadtesting.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/loadtesting.txt)



&#x1F49A; **Microdata Validation**
[microdata.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18509/microdata.txt)

    
--automated message

GHaberis commented on static/src/javascripts/projects/common/views/breaking-news.html:
```diff
@@ -6,10 +6,10 @@
         </div>
     </div>
-    <a class="u-faux-block-link__overlay" href="/<%=id%>"></a>
+    <a href="/<%=id%>"></a>
```
I believe we need to wrap the button in an `<Object>` here and the `<a>` needs to overlay the contents of this container as it did when it had the `.u-faux-block-link__overlay` class as now it'll have no height/width.

GHaberis replied:
@SiAdcock and me have spoken about this already so the comment above is a note to remind us! ☝️

sndrs replied:
shouldn't the link wrap everything, rather than overlay?

SiAdcock commented on static/src/javascripts/projects/common/views/membership-message.html:
```diff
@@ -6,5 +6,5 @@
         </span>
     </div>
-    <a class="u-faux-block-link__overlay js-engagement-message-link" target="_blank" href="<%=linkHref%>" data-link-name="Read more link"></a>
+    <a class="js-engagement-message-link" target="_blank" href="<%=linkHref%>" data-link-name="Read more link"></a>
```
I believe this change will break the (already broken) "do not adblock" message banner

sndrs commented on static/src/stylesheets/base/_helpers.scss:
```diff
@@ -268,3 +268,21 @@
+ * http://kizu.ru/en/fun/nested-links/
+ */
+
+.u-nested-link {
```
`[type=nested/link]`?

sndrs commented on static/src/stylesheets/base/_helpers.scss:
```diff
@@ -268,3 +268,21 @@
+ */
+
+.u-nested-link {
+    width: 0;
```
should these `display: none` in old-ie?

SiAdcock replied:
Since CSS doesn't seem to load (or get applied) at all in <IE9, I don't think there's much point in worrying about it?

sndrs commented on common/app/views/fragments/items/elements/facia_cards/kicker.scala.html:
```diff
@@ -7,7 +7,8 @@
 @makeLink(uri: String, inner: String) = {
-    <a href="@LinkTo(uri)" data-link-name="kicker" class="fc-item__kicker @classes.mkString(" ")">@Html(inner)</a>
+    @* Support nested links by wrapping inner links in <object> element *@
+    <object type="nested/link" class="u-nested-link"><a href="@LinkTo(uri)" data-link-name="kicker" class="fc-item__kicker @classes.mkString(" ")">@Html(inner)</a></object>
```
i wonder if we should extract these into a `nestedLink` helper, and add the IE conditional comments? have a feeling it will totally bork < IE9, and it should at least _work_ there, even it not well?

sndrs replied:
helper may be overkill though

SiAdcock replied:
Yeah, IE9 looks terrible even now... the CSS doesn't seem to load, I'm guess due to some JS compatibility issue. Not sure we should even factor in IE9 support these days, the usage is around 0.27% (incidentally higher than both IE9 and IE10). This can be a question for another time.

I'll add the conditional comments that will kill sublinks on IE9. I think extracting into a helper might be an exercise of DRYness gone too far, as we'd have to find a way of passing all possible attributes to the nested `<a>` or `<button>`. I'm not sure the eventual helper function would be any easier to use than copy-pasting an existing nested link?

If we could do something like:

```js
// helper.js

export const NestedLink = ({ children }) => (
  <object type="nested/link">
    {children}
  </object>
)
```

...then we'd be in business!

SiAdcock commented:
Hi @GHaberis, @sndrs, I have addressed your comments and I believe this is now ready to merge.

Can you take another look, specifically at:

1. [`f20a117`](https://github.com/guardian/frontend/pull/18509/commits/f20a117c34c8e68ae94cd51036e4d162a207185b) Hiding sublinks
2. [`1d43fbd`](https://github.com/guardian/frontend/pull/18509/commits/1d43fbdae814e2b1122fa21234d465a0414c2df5) Fixing breaking news banner

As you know, I'm off until next year. Would be great to get this merged as it solves a few issues and addresses some of the complexity we've been facing around Facia cards.

SiAdcock commented:
This is too big / risky to merge right now. I'll close this, break it down and pick it up again One Day™️
